flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/AbstractArrowPythonAggregateFunctionOperator.java

```diff
@@ -0,0 +1,197 @@
+import org.apache.flink.fnexecution.v1.FlinkFnApi;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.table.api.TableConfig;
+import org.apache.flink.table.data.JoinedRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.binary.BinaryRowData;
+import org.apache.flink.table.functions.AggregateFunction;
+import org.apache.flink.table.functions.python.PythonEnv;
+import org.apache.flink.table.functions.python.PythonFunctionInfo;
+import org.apache.flink.table.planner.codegen.CodeGeneratorContext;
+import org.apache.flink.table.planner.codegen.ProjectionCodeGenerator;
+import org.apache.flink.table.runtime.arrow.serializers.ArrowSerializer;
+import org.apache.flink.table.runtime.arrow.serializers.RowDataArrowSerializer;
+import org.apache.flink.table.runtime.generated.GeneratedProjection;
+import org.apache.flink.table.runtime.generated.Projection;
+import org.apache.flink.table.runtime.operators.python.AbstractStatelessFunctionOperator;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.util.Preconditions;
+
+/**
+ * The Abstract class of Arrow Aggregate Operator for Pandas {@link AggregateFunction}.
+ */
+@Internal
+public abstract class AbstractArrowPythonAggregateFunctionOperator
+	extends AbstractStatelessFunctionOperator<RowData, RowData, RowData> {
+
+	private static final long serialVersionUID = 1L;
+
+	private static final String SCHEMA_ARROW_CODER_URN = "flink:coder:schema:arrow:v1";
+
+	private static final String PANDAS_AGGREGATE_FUNCTION_URN = "flink:transform:pandas_aggregate_function:v1";
+
+	/**
+	 * The Pandas {@link AggregateFunction}s to be executed.
+	 */
+	private final PythonFunctionInfo[] pandasAggFunctions;
+
+	protected final int[] groupingSet;
+
+	protected transient ArrowSerializer<RowData> arrowSerializer;
+
+	/**
+	 * The collector used to collect records.
+	 */
+	protected transient StreamRecordRowDataWrappingCollector rowDataWrapper;
+
+	/**
+	 * The JoinedRowData reused holding the execution result.
+	 */
+	protected transient JoinedRowData reuseJoinedRow;
+
+	/**
+	 * The current number of elements to be included in an arrow batch.
+	 */
+	protected transient int currentBatchCount;
+
+	/**
+	 * The Projection which projects the udaf input fields from the input row.
+	 */
+	private transient Projection<RowData, BinaryRowData> udafInputProjection;
+
+	public AbstractArrowPythonAggregateFunctionOperator(
+		Configuration config,
+		PythonFunctionInfo[] pandasAggFunctions,
+		RowType inputType,
+		RowType outputType,
+		int[] groupingSet,
+		int[] udafInputOffsets) {
+		super(config, inputType, outputType, udafInputOffsets);
+		this.pandasAggFunctions = Preconditions.checkNotNull(pandasAggFunctions);
+		this.groupingSet = Preconditions.checkNotNull(groupingSet);
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
```

Comment:
Should we disable the timer based **finishBundle**?

Reply:
I think we still need the timer based finishBundle. We can still use elementCount and time to control the frequency of sending data in terms of quantity and time